### DISCUSSION TITLE
rbenv install 2.6.5 failing: BUILD FAILED (macOS 11.6 using ruby-build 20210928)

### DISCUSSION THREAD

**Discussion by amitmahajan12** [rbenv/ruby-build#2221] [Build failures]
Please see the attached log. It is unclear to me what is wrong. I'll appreciate any help I can get.

% rbenv install 2.6.5
Downloading openssl-1.1.1l.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/0b7a3e5e59c34827fe0c3a74b7ec8baef302b98fa80088d7f9153aa16fa76bd1
Installing openssl-1.1.1l...
Installed openssl-1.1.1l to /Users/amitmahajan/.rbenv/versions/2.6.5

Downloading ruby-2.6.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.5.tar.bz2
Installing ruby-2.6.5...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 11.6 using ruby-build 20210928)

Inspect or clean up the working tree at /var/folders/qn/6glckh8s1jn5x74lj7bkm7gr0000gn/T/ruby-build.20211006144832.95659.r8FA33
Results logged to /var/folders/qn/6glckh8s1jn5x74lj7bkm7gr0000gn/T/ruby-build.20211006144832.95659.log

Last 10 log lines:
        imemo_type_ids[9] = rb_intern("imemo_parser_strterm");
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../.././include/ruby/ruby.h:1826:56: note: expanded from macro 'rb_intern'
        __extension__ (RUBY_CONST_ID_CACHE((ID), (str))) : \
                                                       ^
1340 warnings generated.
linking shared-object objspace.bundle
422 warnings generated.
linking shared-object date_core.bundle
make: *** [build-ext] Error 2

[ruby-build.20211006144832.95659.log](https://github.com/rbenv/rbenv/files/7297393/ruby-build.20211006144832.95659.log)

**Comment by me25hage** [comment]
I have this same Build Failed error with 3.0.0 except, my build fails during the install of openssl-1.1.1l. Log attached.

[ruby-build.20211012143529.16821.log](https://github.com/rbenv/ruby-build/files/7332452/ruby-build.20211012143529.16821.log)

Downloading openssl-1.1.1l.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/0b7a3e5e59c34827fe0c3a74b7ec8baef302b98fa80088d7f9153aa16fa76bd1
Installing openssl-1.1.1l...

BUILD FAILED (macOS 11.6 using ruby-build 20210928)

Inspect or clean up the working tree at /var/folders/h8/y2qk65cs3zd35cq2frzj_wqc0000gn/T/ruby-build.20211012143529.16821.OcOFOF
Results logged to /var/folders/h8/y2qk65cs3zd35cq2frzj_wqc0000gn/T/ruby-build.20211012143529.16821.log

Last 10 log lines:
${LDCMD:-clang} -arch arm64 -O3 -Wall -L. -Wl,-search_paths_first -L/usr/local/var/rbenv/versions/3.0.0/lib  \
		-o fuzz/client-test fuzz/client.o fuzz/test-corpus.o \
		 -lssl -lcrypto  
ld: warning: directory not found for option '-L/usr/local/var/rbenv/versions/3.0.0/lib'
ld: warning: directory not found for option '-L/usr/local/var/rbenv/versions/3.0.0/lib'
ld: warning: directory not found for option '-L/usr/local/var/rbenv/versions/3.0.0/lib'
make depend && make _build_libs
make[1]: Nothing to be done for `_build_libs'.
Cannot create directory /usr/local/var: Permission denied
make: *** [install_runtime_libs] Error 2

**Comment by amitmahajan12** [comment]
I had multiple versions of ruby installed. After I uninstalled everything and reinstalled only 2.6.5, it worked for me. I think failing during install of openssl-1.1.1l may be a different issue anyway.

**Comment by gplusdotgr** [comment]
```
v14.18.1  ~/r/documentation    rbenv install 2.6.4                                                  Mon Nov 29 12:06:37 2021
Downloading openssl-1.1.1l.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/0b7a3e5e59c34827fe0c3a74b7ec8baef302b98fa80088d7f9153aa16fa76bd1
Installing openssl-1.1.1l...
Installed openssl-1.1.1l to /Users/gkatsanos/.rbenv/versions/2.6.4

Downloading ruby-2.6.4.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.4.tar.bz2
Installing ruby-2.6.4...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 12.0.1 using ruby-build 20211109)

Inspect or clean up the working tree at /var/folders/y5/dphtsvm95_38xml5cs344hn80000gn/T/ruby-build.20211129120640.19263.BPu93D
Results logged to /var/folders/y5/dphtsvm95_38xml5cs344hn80000gn/T/ruby-build.20211129120640.19263.log

Last 10 log lines:
linking shared-object stringio.bundle
linking shared-object -test-/string.bundle
5 warnings generated.
4 warnings generated.
linking shared-object zlib.bundle
1 warning generated.
linking shared-object psych.bundle
422 warnings generated.
linking shared-object date_core.bundle
make: *** [build-ext] Error 2
```

nothing works ..

**Comment by rivitt** [comment]
I'm getting this same error on macOS 11.6 and 12.0.1 (m1 chip), I'm going to uninstall and reinstall everything and see where that takes me

**Reply by HMADAHMD** [reply]
have you got any solution?

**Reply by mrmeszaros** [reply]
Hi, in short, generally do not try to install old patch versions of ruby.
So instead of using 2.6.5, install the latest, like (currently) 2.6.10

If You want a long, more in depth explanation, read this answer: https://github.com/rbenv/ruby-build/discussions/2221#discussioncomment-6570276

**Comment by iamricks** [comment]
I'm getting the same issue trying to install 2.6.3 but i am able to install and use 3.0.2 and 3.0.3 just fine, maybe the issue is just with 2.6.x?

**Comment by x-warukira** [comment]
`export optflags="-Wno-error=implicit-function-declaration"` worked for me

**Reply by odravison** [reply]
What does it?

**Reply by odravison** [reply]
Also, it worked for me xD

**Comment by JShefferCorp** [comment]
> export optflags="-Wno-error=implicit-function-declaration"

Thanks, this also worked for me.  
MacOS: 12.5
Processor: Apple M1 Pro

**Comment by markst** [comment]
Legend @x-warukira! That's finally fixed being unable to install `2.6.0` on my M1

**Comment by mrmeszaros** [comment]
> `export optflags="-Wno-error=implicit-function-declaration"` worked for me

This Worked on my machine
MacOS: `Ventura (13.4.1)`
Processor: `Apple M1 Pro`

On that note - would it be possible to persistently set that variable for the context of `ruby-build` or `rbenv-install`?

@x-warukira How did You find the solution? I am interested to see what component reads the `optflags` variable!

**Comment by monfresh** [comment]
There are several misconceptions going on here. First, the "-Wno-error=implicit-function-declaration" flag should never be set. It's a dangerous hack that overrides macOS defaults, as explained here: https://github.com/rbenv/ruby-build/issues/2001#issuecomment-1181613673

I've been using Ruby for 13 years and not a single time have I ever had to set that flag. If you have a proper Ruby dev setup, and you're installing the appropriate version of Ruby in the appropriate context, you should never need to set that flag.

The second misconception is that if your project is using a specific version of Ruby, then your only option is to install that specific version. I see so many people try to install older versions of Ruby 2.6.x, like in this GitHub issue, when the solution is to update your project to the latest patch version (the 3rd digit), which is 2.6.10.  

rbenv/ruby-build can install 2.6.x versions as old as 2.6.8 on Apple Silicon Macs in native mode without any issues and without having to set any flags.

Here's a detailed article I wrote that explains [how and why to upgrade the Ruby version in your project](https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project).

If for some highly unlikely reason you absolutely need to use a version of 2.6.x older than 2.6.8, then the best option is to use Rosetta.

**Reply by mrmeszaros** [reply]
Thanks for the clarification!
As You said - the latest 2.7.x release was installed without issue.
I wanted to have 2.7.0 to mirror  machine in production, but hopefully this won't cause any issue.

**Comment by azisramdhan** [comment]
> export optflags="-Wno-error=implicit-function-declaration"

Thank you, it works for me as well

MacOS 14.5
Apple M1
